eZ/Publish/Core/FieldType/Keyword/KeywordStorage/Gateway/LegacyStorage.php

```diff
@@ -286,12 +287,11 @@ protected function insertKeywords(array $keywordsToInsert, $contentTypeId)
         return $keywordIdMap;
     }
 
-    protected function deleteOldKeywordAssignments($fieldId, VersionInfo $versionInfo = null)
+    protected function deleteOldKeywordAssignments($fieldId, VersionInfo $versionInfo)
     {
         $dbHandler = $this->getConnection();
         // do not delete assignments if there's other published version containing this field
-        if ($versionInfo !== null &&
-            $versionInfo->status !== VersionInfo::STATUS_PUBLISHED &&
+        if ($versionInfo->status === VersionInfo::STATUS_ARCHIVED &&
```

Comment:
> But then update comment to say something like
`// If current version being asked to be deleted is archive, then don't do that if there is some other version which is published (as keyword table is not versioned)`

Added in 0a442f4

> But it does bring the question how this behaves when deleting a draft, does it delete keywords of the published content?

Uhm, in that case it [fails](https://travis-ci.org/ezsystems/ezpublish-kernel/jobs/210502435#L696) - added this test in dc983f2.

I tried [here](https://github.com/alongosz/ezpublish-kernel/blob/0a442f4cda356f4ef26c58fb05557ae645d00658/eZ/Publish/Core/FieldType/Keyword/KeywordStorage/Gateway/LegacyStorage.php#L295-L296) different checks for `status`, `contentInfo->isPublished`, `versionNo` and `contentInfo->currentVersionNo` and I cannot see at this point any solution that covers all the cases :(

If I check for non-published version while content is published, then other tests fail, if I check for archived version while content is published then deleting draft deletes keywords.